### ISSUE TITLE
Publish roadmap

### ISSUE THREAD

**Issue by vlsi** [javacc/javacc#239]
As per [the comment](https://github.com/javacc/javacc-8/discussions/2#discussioncomment-3216593), none of the features can be merged to core (I don't understand what core is) without @new-javacc approval.

However, every time I ask(https://github.com/javacc/javacc/issues/237, https://github.com/javacc/javacc/issues/236, https://github.com/javacc/javacc/issues/235, https://github.com/javacc/javacc/issues/234), new-javacc says they are not interested.

Is there a roadmap or a list of in-progress features so I could better understand the evolution of the project?

**Comment by opeongo** [comment]
@vlsi you are missing the point.  There is no code to be 'merged'.  None of the features that you have mentioned can be 'merged' until someone writes the code.  @new-javacc has stated multiple times that he is not available or not interested enough to write the code for these features that you have mentioned.  There are no employees here that will write this code for you,   If you are interested in these feature then go ahead and write the code yourself and submit it for review.  That is generally how open source works.

Javacc21 is a different project than javacc.  It is not a simple case of doing a git merge command from one code base to another.

**Comment by vlsi** [comment]
@opeongo , I am afraid you miss the point.

1) Here's the comment which says only Sreeni is ever allowed to modify "core": https://github.com/javacc/javacc-8/discussions/2#discussioncomment-3216593

2) The key question here is to get idea if javacc/javacc has anything on their roadmap.

3) "not like open source works".
There's a trivial pull request https://github.com/javacc/javacc/pull/85 waiting for more than 3 years.
I am pretty sure it means bigger changes will never be reviewed/merged.

**Comment by revusky** [comment]
> Javacc21 is a different project than javacc.

Well, it is true that JavaCC21 is a different project but that's not really a fully correct characterization of the situation. JavaCC21 is a result of a continuation of work on the original JavaCC that was open-sourced by Sun Microsystems in June 2003. That branch of development dates back to 2008, when it was originally called FreeCC, and it was resuscitated as JavaCC21 in early 2020. See https://doku.javacc.com/doku.php?id=ancient_history

The reason that FreeCC (as it was called at the time) came into existence was simply because the people who controlled the canonical or legacy project (Sreeni and his acolytes) were simply impossible. They had no roadmap for further development of the codebase (still don't) and simply rejected (with no real discussion) any new ideas.

Now, FreeCC/JavaCC21 (and soon to be renamed CongoCC, BTW) is not a "fork" either. A "fork" (usually considered something to avoid if possible) is a *bifurcation* of effort. It means that one group takes one path and the other takes another path. However, there is no such *bifurcation* because what we have here is simply that there is one branch of ongoing development and another group that simply never does anything. That is not a "bifurcation". 

> That is generally how open source works.

Uhh, no. It's how open source doesn't work! LOL!

**Comment by opeongo** [comment]
@revusky what is the backwards compatibility between Javacc21 and this javacc project?  Is javacc21 a drop in replacement for an existing jj or jjt file, or is there a different command set or syntax?

**Comment by revusky** [comment]
> @revusky what is the backwards compatibility between Javacc21 and this javacc project? Is javacc21 a drop in replacement for an existing jj or jjt file, or is there a different command set or syntax?

In terms of generating a parser from a legacy jj or jjt file, you could say it is a drop-in replacement. Pretty much.  That is because legacy syntax continues to be supported. It will *probably* generate the parser code from your grammar with no fuss, though you should try it, of course.

However, that does not mean that you can just switch to JavaCC21 with no adjustment. This is because the client code that uses the generated parser would need adjustment. If you're wondering why, you could read this: https://javacc.com/2021/12/05/tokens-then-and-now/
and the following blog post: https://javacc.com/2021/12/06/token-chaining-javaccs-dark-underbelly/

Basically, the only way to fix issues in legacy JavaCC involves non-backward-compatible changes to the generated API. That is because the legacy JavaCC generates code with directly accessible public fields in Token.java and any client code that uses that, i.e. tok.beginLine or tok.image would have to be ajusted. There was no way to keep those things working because the fields in question were refactored out of existence, so at the very least, you'd have to replace `tok.image` with `tok.getImage()`. Things like that. Generally speaking, your client code would have to be adjusted to reflect changes in the generated API. The upside to that is that the generated API is really much better, uses modern Java idioms etcetera.
